src/2025h2/evolving-traits.md

```diff
@@ -0,0 +1,266 @@
+
+### The status quo
+
+Today's solutions include:
+
+### Add a supertrait
+
+#### Pros
+* Matches the conceptual pattern between the traits: one clearly implies the other.
+* The "standard" Rust way of doing things.
+
+#### Cons
+* Breaking change.
+* Requires implementors to split out the implementation into separate `impl` blocks for each trait.
+
+### Add a blanket impl
+
+Rather than declaring `trait A: B`, one can create sibling traits with a blanket impl:
+
+```rust
+trait Supertrait {
+  // supertrait items
+}
+
+trait Subtrait {
+  // supertrait items + subtrait items
+}
+
+impl<T: Subtrait> Supertrait for T {
+  // impl supertrait items using subtrait items
+}
+```
+
+#### Pros
+* Backwards compatible to introduce `Supertrait`
+
+#### Cons
+
+* Middleware impls are impossible! We'd like to write:
+
+```rust
+struct Middeware<T>(T);
+
+impl<T: Supertrait> Supertrait for Middleware<T> { ... }
+
+impl<T: Subtrait> Subtrait for Middleware<T> { ... }
+```
+
+but this overlaps with the blanket impl, and is rejected by the Rust compiler! This is a critical issue for `async` bridge APIs such as tower's `Service` trait, which wants to provide wrappers which implement the `trait_variant`-style `Send`-able when the underlying type implements the `Send` version (see [these notes from a previous design meeting](https://hackmd.io/rmN25qziSHKT4kv-ZC8QPw)).
+
+Other nits:
+* The directionality of the impl is less clear.
+* Every shared item has two names: `<T as Supertrait>::Item` and `<T as Subtrait>::Item`. Relatedly, the bridge impl must exist even if items are identical.
+
+### Provide no bridging
+
+Another alternative is to provide two totally separate traits with no bridging, requiring users to manually implement both versions of the trait:
+
+```rust
+trait Supertrait { ... }
+trait Subtrait { ... }
+
+struct Impl { ... }
+
+impl Supertrait for T { ... }
+impl Subtrait for T { ... }
+```
+
+#### Pros
+
+* Backwards compatible
+* Middleware can be written to bridge either impl
+
+#### Cons
+
+* Requires duplication
+* Requires users to restate bounds
+* Existing code which implements `Subtrait` cannot be used as `Supertrait`, so APIs which require `Subtrait` cannot be relaxed to `Supertrait`
+
+### The next 6 months
+
+In the next six months, we aim to ship a solution which addresses the `Service` and `Receiver` use-cases
+by allowing trait impls to implement supertraits *if* the impl itself contains a definition of an item
+from the supertrait.
+
+Traits will have to opt their impls into this behavior, possibly through the use of a keyword.
+`auto` is used below as an example keyword:
+
+```rust
+// Library code:
+trait Subtrait {
+    fn supertrait_item();
+    fn subtrait_item();
+}
+// User code:
+impl Subtrait for MyType {
+    fn supertrait_item() { ... }
+    fn subtrait_item() { ... }
+}
+
+// -- can become --
+
+// Library code:
+trait Supertrait {
+    fn supertrait_item();
+}
+trait Subtrait: Supertrait {
```

Comment:
Out of curiosity, are you thinking of leaving out the need to write `auto impl Supertrait` or similar when the supertrait has unprovided items?

Reply:
We can't require the impl itself to use a keyword, as this would make splitting up a trait a blocking change (impls would now be required to add a keyword they didn't previously use).